File: ViewModels/MondayItemDetail/MondayItemDetailBag.cs
using com.baysideonline.BccMonday.Utilities.Api.Interfaces;
using com.baysideonline.BccMonday.Utilities.Api.Schema;
using Rock.Web.UI.Controls;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace com.baysideonline.BccMonday.ViewModels.MondayItemDetail
{
    public class MondayItemDetailBag
    {
        public MondayItemBag Item { get; set; }

        public string Status { get; set; }
        public int StatusIndex { get; set; }

        public bool ShowApprove { get; set; }
        public bool ShowClose { get; set; }
    }

    public class MondayItemBag
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string CreatedAt { get; set; }
        public MondayBoardBag Board { get; set; }
        public List<MondayUpdateBag> Updates { get; set; }
        public List<MondayColumnValueBag> ColumnValues { get; set; }
    }

    public class MondayBoardBag
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }

    public class MondayUpdateBag
    {
        public string Id { get; set; }
        public string CreatedAt { get; set; }
        public string CreatorName { get; set; }
        public string TextBody { get; set; }
        public List<MondayUpdateBag> Replies { get; set; }
        public List<MondayAssetBag> Files { get; set; }
    }

    public class MondayAssetBag
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string PublicUrl { get; set; }
    }

    public class MondayColumnBag
    {
        public string Id { get; set; }
        public string Title { get; set; }
    }

    public class MondayColumnValueBag
    {
        public string Id { get; set; }
        public string Text { get; set; }
        public string Type { get; set; }
        public string Value { get; set; }
        public MondayColumnBag Column { get; set; }

        //Properties specific to FileColumnValue
        public List<FileBag> Files { get; set; }

        //Properties specific to StatusColumnValue
        public int? Index { get; set; }
        public string StatusLabel { get; set; }
        public bool? IsDone { get; set; }
        public LabelStyle LabelStyle { get; set; }

        //Properties specific to BoardRelationColumnValue
        public string DisplayValue { get; set; }
        public List<LinkedItem> LinkedItems { get; set; }
        public List<long> LinkedItemIds { get; set; }
    }

    public class FileBag
    {
        public string AssetId { get; set; }
        public MondayAssetBag Asset { get; set; }
    }

    public class LabelStyle
    {
        public string Color { get; set; }
        public string Border { get; set; }
    }

    public class LinkedItem
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }
}

File: src/ViewModels/MondayItemDetailBag.ts

export type MondayItemDetailBag = {
    item: MondayItemBag
    status: string;
    statusIndex: number;
    showApprove: boolean;
    showClose: boolean;
}

export type MondayItemBag = {
    id: string;
    createdAt: string;
    name: string;
    board: MondayBoardBag;
    columnValues: ColumnValueBag[];
    updates: MondayUpdateBag[];
}

export type MondayBoardBag = {
    id: string;
    name: string;
}

export type MondayUpdateBag = {
    id: string;
    createdAt: string;
    creatorName: string;
    textBody: string;
    replies: MondayUpdateBag[] | null;
    files: MondayAssetBag[] | null
}

export type MondayAssetBag = {
    id: number | string;
    name: string;
    publicUrl: string;
}

export type ColumnValueBag = {
    id: number | string;
    text: string;
    type: keyof typeof ColumnType;
    value: string;
    column: Column;
    files?: {
        assetId: string;
        asset: {
            url: string;
            name: string;
            urlThumbnail: string;
            publicUrl: string;
        };
    }[];
    labelStyle?: {
        color: string;
        border: string;
    };
    linkedItemIds?: string[]
}

export type BasicColumnValue = {
    id: number | string;
    text: string;
    type: keyof typeof ColumnType;
    value: string;
    column: Column
}

export type Column = {
    id: string;
    title: string;
}

export type FileColumnValue = BasicColumnValue & {
    files: {
        assetId: string;
        asset: {
            url: string;
            name: string;
            urlThumbnail: string;
            publicUrl: string;
        };
    }[]
}

export type StatusColumnValue = BasicColumnValue & {
    index: number;
    statusLabel: string;
    isDone: boolean;
    labelStyle: {
        color: string;
        border: string;
    }
}

export type BoardRelationColumnValue = BasicColumnValue & {
    displayValue: string;
    linkedItems: {
        id: string;
        name: string;
    }
    linkedItemIds: string[]
}

export type AbstractColumnValue =
    BoardRelationColumnValue
    | FileColumnValue
    | StatusColumnValue
    | BasicColumnValue

export enum ColumnType {
    auto_number,
    board_relation,
    button,
    checkbox,
    color_picker,
    country,
    creation_log,
    date,
    dependency,
    doc,
    dropdown,
    email,
    file,
    formula,
    hour,
    item_assignees,
    item_id,
    last_updated,
    link,
    location,
    long_text,
    mirror,
    name,
    numbers,
    people,
    phone,
    progress,
    rating,
    status,
    subtasks,
    tags,
    team,
    text,
    timeline,
    time_tracking,
    vote,
    week,
    world_clock,
    unsupported,
}